File: templates/Uwp/Pages/InkDraw/ViewModels/InkDrawViewViewModel.cs
using System.Windows.Input;
using Param_ItemNamespace.Services.Ink;
using Param_ItemNamespace.Helpers;

namespace Param_ItemNamespace.ViewModels
{
    public class InkDrawViewViewModel : System.ComponentModel.INotifyPropertyChanged
    {
        private InkStrokesService _strokeService;
        private InkLassoSelectionService _lassoSelectionService;
        private InkPointerDeviceService _pointerDeviceService;
        private InkCopyPasteService _copyPasteService;
        private InkUndoRedoService _undoRedoService;
        private InkFileService _fileService;
        private InkZoomService _zoomService;

        private ICommand cutCommand;
        private ICommand copyCommand;
        private ICommand pasteCommand;
        private ICommand undoCommand;
        private ICommand redoCommand;
        private ICommand zoomInCommand;
        private ICommand zoomOutCommand;
        private ICommand loadInkFileCommand;
        private ICommand saveInkFileCommand;
        private ICommand exportAsImageCommand;
        private ICommand clearAllCommand;

        private bool enableTouch = true;
        private bool enableMouse = true;
        private bool enableLassoSelection;

        public InkDrawViewViewModel()
        {
        }

        public void Initialize(
            InkStrokesService strokeService,
            InkLassoSelectionService lassoSelectionService,
            InkPointerDeviceService pointerDeviceService,
            InkCopyPasteService copyPasteService,
            InkUndoRedoService undoRedoService,
            InkFileService fileService,
            InkZoomService zoomService)
        {
            _strokeService = strokeService;
            _lassoSelectionService = lassoSelectionService;
            _pointerDeviceService = pointerDeviceService;
            _copyPasteService = copyPasteService;
            _undoRedoService = undoRedoService;
            _fileService = fileService;
            _zoomService = zoomService;

            _pointerDeviceService.DetectPenEvent += (s, e) => EnableTouch = false;
        }

        public ICommand CutCommand => cutCommand
           ?? (cutCommand = new RelayCommand(() =>
           {
               _copyPasteService?.Cut();
               ClearSelection();
           }));

        public ICommand CopyCommand => copyCommand
           ?? (copyCommand = new RelayCommand(() => _copyPasteService?.Copy()));

        public ICommand PasteCommand => pasteCommand
           ?? (pasteCommand = new RelayCommand(() =>
           {
               _copyPasteService?.Paste();
               ClearSelection();
           }));

        public ICommand UndoCommand => undoCommand
           ?? (undoCommand = new RelayCommand(() =>
           {
               ClearSelection();
               _undoRedoService?.Undo();
           }));

        public ICommand RedoCommand => redoCommand
           ?? (redoCommand = new RelayCommand(() =>
           {
               ClearSelection();
               _undoRedoService?.Redo();
           }));

        public ICommand ZoomInCommand => zoomInCommand
            ?? (zoomInCommand = new RelayCommand(() => _zoomService?.ZoomIn()));

        public ICommand ZoomOutCommand => zoomOutCommand
            ?? (zoomOutCommand = new RelayCommand(() => _zoomService?.ZoomOut()));

        public ICommand LoadInkFileCommand => loadInkFileCommand
           ?? (loadInkFileCommand = new RelayCommand(async () =>
           {
               ClearSelection();
               var fileLoaded = await _fileService?.LoadInkAsync();

               if (fileLoaded)
               {
                   _undoRedoService?.Reset();
               }
           }));

        public ICommand SaveInkFileCommand => saveInkFileCommand
           ?? (saveInkFileCommand = new RelayCommand(async () =>
           {
               ClearSelection();
               await _fileService?.SaveInkAsync();
           }));

        public ICommand ExportAsImageCommand => exportAsImageCommand
           ?? (exportAsImageCommand = new RelayCommand(async () =>
           {
               ClearSelection();
               await _fileService?.ExportToImageAsync();
           }));

        public ICommand ClearAllCommand => clearAllCommand
           ?? (clearAllCommand = new RelayCommand(ClearAll));

        public bool EnableTouch
        {
            get => enableTouch;
            set
            {
                Param_Setter(ref enableTouch, value);
                _pointerDeviceService.EnableTouch = value;
            }
        }

        public bool EnableMouse
        {
            get => enableMouse;
            set
            {
                Param_Setter(ref enableMouse, value);
                _pointerDeviceService.EnableMouse = value;
            }
        }

        public bool EnableLassoSelection
        {
            get => enableLassoSelection;
            set
            {
                Param_Setter(ref enableLassoSelection, value);
                ConfigLassoSelection(value);
            }
        }

        private void ConfigLassoSelection(bool enableLasso)
        {
            if (enableLasso)
            {
                _lassoSelectionService?.StartLassoSelectionConfig();
            }
            else
            {
                _lassoSelectionService?.EndLassoSelectionConfig();
            }
        }

        private void ClearAll()
        {
            ClearSelection();
            _strokeService?.ClearStrokes();
            _undoRedoService?.Reset();
        }

        private void ClearSelection() => _lassoSelectionService?.ClearSelection();
    }
}

File: templates/Uwp/Pages/InkDrawVB/ViewModels/InkDrawViewViewModel.vb
Imports System.Windows.Input
Imports Param_ItemNamespace.Services.Ink
Imports Param_ItemNamespace.Helpers

Namespace ViewModels
    Public Class InkDrawViewViewModel
        Inherits System.ComponentModel.INotifyPropertyChanged

        Private _strokeService As InkStrokesService
        Private _lassoSelectionService As InkLassoSelectionService
        Private _pointerDeviceService As InkPointerDeviceService
        Private _copyPasteService As InkCopyPasteService
        Private _undoRedoService As InkUndoRedoService
        Private _fileService As InkFileService
        Private _zoomService As InkZoomService
        Private _cutCommand As ICommand
        Private _copyCommand As ICommand
        Private _pasteCommand As ICommand
        Private _undoCommand As ICommand
        Private _redoCommand As ICommand
        Private _zoomInCommand As ICommand
        Private _zoomOutCommand As ICommand
        Private _loadInkFileCommand As ICommand
        Private _saveInkFileCommand As ICommand
        Private _exportAsImageCommand As ICommand
        Private _clearAllCommand As ICommand
        Private _enableTouch As Boolean = True
        Private _enableMouse As Boolean = True
        Private _enableLassoSelection As Boolean

        Public Sub New()
        End Sub

        Public Sub Initialize(strokeService As InkStrokesService, lassoSelectionService As InkLassoSelectionService, pointerDeviceService As InkPointerDeviceService, copyPasteService As InkCopyPasteService, undoRedoService As InkUndoRedoService, fileService As InkFileService, zoomService As InkZoomService)
            _strokeService = strokeService
            _lassoSelectionService = lassoSelectionService
            _pointerDeviceService = pointerDeviceService
            _copyPasteService = copyPasteService
            _undoRedoService = undoRedoService
            _fileService = fileService
            _zoomService = zoomService
            AddHandler _pointerDeviceService.DetectPenEvent, Sub(s, e) EnableTouch = False
        End Sub

        Public ReadOnly Property CutCommand As ICommand
            Get
                If _cutCommand Is Nothing Then
                    _cutCommand = New RelayCommand(Sub()
                                                       _copyPasteService?.Cut()
                                                       ClearSelection()
                                                   End Sub)
                End If

                Return _cutCommand
            End Get
        End Property

        Public ReadOnly Property CopyCommand As ICommand
            Get
                If _copyCommand Is Nothing
                    _copyCommand = New RelayCommand(Sub() _copyPasteService?.Copy())

                End If

                Return _copyCommand
            End Get
        End Property

        Public ReadOnly Property PasteCommand As ICommand
            Get
                If _pasteCommand Is Nothing Then
                    _pasteCommand = New RelayCommand(Sub()
                                                         _copyPasteService?.Paste()
                                                         ClearSelection()
                                                     End Sub)
                End If

                Return _pasteCommand
            End Get
        End Property

        Public ReadOnly Property UndoCommand As ICommand
            Get
                If _undoCommand Is Nothing Then
                    _undoCommand = New RelayCommand(Sub()
                                                        ClearSelection()
                                                        _undoRedoService?.Undo()
                                                    End Sub)
                End If

                Return _undoCommand
            End Get
        End Property

        Public ReadOnly Property RedoCommand As ICommand
            Get
                If _redoCommand Is Nothing Then
                    _redoCommand = New RelayCommand(Sub()
                                                        ClearSelection()
                                                        _undoRedoService?.Redo()
                                                    End Sub)
                End If

                Return _redoCommand
            End Get
        End Property

        Public ReadOnly Property ZoomInCommand As ICommand
            Get
                If _zoomInCommand Is Nothing Then
                    _zoomInCommand = New RelayCommand(Sub() _zoomService?.ZoomIn())
                End If

                Return _zoomInCommand
            End Get
        End Property

        Public ReadOnly Property ZoomOutCommand As ICommand
            Get
                If _zoomOutCommand Is Nothing Then
                    _zoomOutCommand = New RelayCommand(Sub() _zoomService?.ZoomOut())
                End If

                Return _zoomOutCommand
            End Get
        End Property

        Public ReadOnly Property LoadInkFileCommand As ICommand
            Get
                If _loadInkFileCommand Is Nothing Then
                    _loadInkFileCommand = New RelayCommand(Async Sub()
                                                               ClearSelection()
                                                               Dim fileLoaded = Await _fileService?.LoadInkAsync()

                                                               If fileLoaded Then
                                                                   _undoRedoService?.Reset()
                                                               End If
                                                           End Sub)
                End If

                Return _loadInkFileCommand
            End Get
        End Property

        Public ReadOnly Property SaveInkFileCommand As ICommand
            Get
                If _saveInkFileCommand Is Nothing Then
                    _saveInkFileCommand = New RelayCommand(Async Sub()
                                                               ClearSelection()
                                                               Await _fileService?.SaveInkAsync()
                                                           End Sub)
                End If

                Return _saveInkFileCommand
            End Get
        End Property

        Public ReadOnly Property ExportAsImageCommand As ICommand
            Get
                If _exportAsImageCommand Is Nothing Then
                    _exportAsImageCommand = New RelayCommand(Async Sub()
                                                                 ClearSelection()
                                                                 Await _fileService?.ExportToImageAsync()
                                                             End Sub)
                End If

                Return _exportAsImageCommand
            End Get
        End Property

        Public ReadOnly Property ClearAllCommand As ICommand
            Get
                If _clearAllCommand Is Nothing Then
                    _clearAllCommand = New RelayCommand(AddressOf ClearAll)
                End If

                Return _clearAllCommand
            End Get
        End Property

        Public Property EnableTouch As Boolean
            Get
                Return _enableTouch
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableTouch, value)
                _pointerDeviceService.EnableTouch = value
            End Set
        End Property

        Public Property EnableMouse As Boolean
            Get
                Return _enableMouse
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableMouse, value)
                _pointerDeviceService.EnableMouse = value
            End Set
        End Property

        Public Property EnableLassoSelection As Boolean
            Get
                Return _enableLassoSelection
            End Get
            Set(value As Boolean)
                [Param_Setter](_enableLassoSelection, value)
                ConfigLassoSelection(value)
            End Set
        End Property

        Private Sub ConfigLassoSelection(enableLasso As Boolean)
            If enableLasso Then
                _lassoSelectionService?.StartLassoSelectionConfig()
            Else
                _lassoSelectionService?.EndLassoSelectionConfig()
            End If
        End Sub

        Private Sub ClearAll()
            ClearSelection()
            _strokeService?.ClearStrokes()
            _undoRedoService?.Reset()
        End Sub

        Private Sub ClearSelection()
            _lassoSelectionService?.ClearSelection()
        End Sub
    End Class
End Namespace
